### PR TITLE
Homme(xx)/SL: Performance-portable semi-Lagrangian transport + physgrid

### PR DESCRIPTION
**Pull request by ambrad**
This PR brings in two major features:
- Semi-Lagrangian tracer transport supports both the F90 theta-l and C++/Kokkos theta-l_kokkos dycores; thus, the second supports SCREAM. The first is at performance parity with master.
- Physgrid remap supports both EAM and SCREAM. Again, the first is at performance parity with master.

A minor new feature is SL transport supports doubly-periodic planar geometry.

This PR also has new tests:
- 6 new standalone Homme tests: ne2 and ne6 theta-l vs theta-l_kokkos with SL;
- 2 new unit tests: gllfvremap_ut for physgrid remap and compose_ut for SL transport.

Finally, this PR makes four unit tests run in standalone Homme optimized builds: gllfvremap_ut, compose_ut, limiters_ut, dirk_ut.

[BFB] for EAM. BFB for standalone Homme except: (1) 8 new standalone Homme tests, (2) thetah-sl-dcmip16_test1pg2 in the BFB build, and (3) all three thetah-sl tests in the optimized build.

### CODE REVIEW

**Comment by ambrad** [comment]
At f529cc2, e3sm_atm_developer and e3sm_atm_integration pass on Chrysalis with midnight 11 Aug master baselines.

**Comment by ambrad** [comment]
At 8618559, the v2 WC PE layout tests are BFB with each other and master.

**Comment by ambrad** [comment]
As one performance comparison, I ran an F-case for 3 months with PE layout 10x16x4 nodes x MPI x threads on Chrysalis and get the timing results listed at the end of this comment. Although in this one case, the branch is faster than master, in general, in various test runs, I find the performance difference statistically indistinguishable, with gaps in either direction depending on node allocation, system load, etc. The "nstep, te" output in these two runs are BFB.
```
>>> master
"CPL:RUN_LOOP"          160 6.912000e+05   3.513364e+05  2195.885 2195.819
"a:CAM_run2"            160 6.913600e+05   6.230462e+04   392.581  385.680
"a:CAM_run3"            160 6.913600e+05   6.731202e+04   424.616  416.369
"a:PAT_remap"           640 2.765440e+06   1.060311e+05   169.135  161.662
"a:SLMM_reconstruct"    640 2.765440e+06   2.357365e+03     4.168    3.310
"a:SLMM_v2x"            640 2.765440e+06   1.405886e+03     2.453    1.901
"a:SLMM_csl"            640 2.765440e+06   4.925961e+04    81.646   72.615
"a:CEDR"                640 2.765440e+06   1.337376e+04    21.328   20.429
"a:CEDR_local"          640 2.765440e+06   1.209680e+04    19.882   18.383
"a:SLMM_bexchV"         640 2.765440e+06   7.381061e+03    13.704    9.427
"a:SLMM vertical remap" 640 2.765440e+06   2.773601e+04    49.586   37.844
"a:CAM_run4"            160 6.913600e+05   4.564321e+02     3.365    2.726
"a:CAM_run1"            160 6.913600e+05   2.056203e+05  1330.641 1215.527
>>> branch
"CPL:RUN_LOOP"          320 1.382400e+06   6.691832e+05  2091.245 2091.175
"a:CAM_run2"            320 1.382720e+06   1.115879e+05   352.069  340.139
"a:CAM_run3"            320 1.382720e+06   1.228855e+05   390.795  375.829
"a:PAT_remap"           640 2.765440e+06   9.901291e+04   158.430  146.987
"a:SLMM_reconstruct"    640 2.765440e+06   2.393273e+03     4.196    3.373
"a:SLMM_v2x"            640 2.765440e+06   1.455064e+03     2.701    1.984
"a:SLMM_csl"            640 2.765440e+06   4.388243e+04    73.383   61.799
"a:CEDR"                640 2.765440e+06   1.260808e+04    20.119   19.104
"a:CEDR_local"          640 2.765440e+06   1.447817e+04    24.215   19.815
"a:SLMM_bexchV"         640 2.765440e+06   6.404439e+03    13.751    7.867
"a:SLMM vertical remap" 640 2.765440e+06   2.774474e+04    48.723   38.218
"a:CAM_run4"            320 1.382720e+06   7.712412e+02     2.995    2.285
"a:CAM_run1"            320 1.382720e+06   4.030191e+05  1299.960 1201.174
```
As an example of performance variability, here's a 1-month 10x32x2 run:
```
>>> master
"CPL:RUN_LOOP"          320 4.761600e+05   2.309410e+05   721.793  721.655
"a:CAM_run2"            320 4.764800e+05   3.814709e+04   120.335  116.376
"a:CAM_run3"            320 4.764800e+05   4.210386e+04   133.828  128.720
"a:PAT_remap"           640 9.529600e+05   3.363302e+04    53.676   50.018
"a:SLMM_reconstruct"    640 9.529600e+05   8.219957e+02     1.444    1.133
"a:SLMM_v2x"            640 9.529600e+05   4.884449e+02     0.841    0.657
"a:SLMM_csl"            640 9.529600e+05   1.544743e+04    25.841   21.711
"a:CEDR"                640 9.529600e+05   4.402231e+03     6.997    6.734
"a:CEDR_local"          640 9.529600e+05   4.116502e+03     6.753    5.728
"a:SLMM_bexchV"         640 9.529600e+05   2.197685e+03     4.635    2.732
"a:SLMM vertical remap" 640 9.529600e+05   9.563318e+03    16.945   13.184
"a:CAM_run4"            320 4.764800e+05   6.327773e+02     2.150    1.931
"a:CAM_run1"            320 4.764800e+05   1.387989e+05   447.777  412.426
>>> branch
"CPL:RUN_LOOP"          320 4.761600e+05   2.333804e+05   729.368  729.268
"a:CAM_run2"            320 4.764800e+05   3.855713e+04   121.705  117.564
"a:CAM_run3"            320 4.764800e+05   4.237204e+04   134.985  129.566
"a:PAT_remap"           640 9.529600e+05   3.418750e+04    54.917   50.813
"a:SLMM_reconstruct"    640 9.529600e+05   8.233109e+02     1.449    1.133
"a:SLMM_v2x"            640 9.529600e+05   5.009489e+02     0.925    0.684
"a:SLMM_csl"            640 9.529600e+05   1.513252e+04    25.253   21.326
"a:CEDR"                640 9.529600e+05   4.343793e+03     6.919    6.625
"a:CEDR_local"          640 9.529600e+05   5.004995e+03     8.409    6.881
"a:SLMM_bexchV"         640 9.529600e+05   2.257322e+03     4.854    2.745
"a:SLMM vertical remap" 640 9.529600e+05   9.564083e+03    16.951   13.215
"a:CAM_run4"            320 4.764800e+05   1.756487e+03     5.917    5.442
"a:CAM_run1"            320 4.764800e+05   1.392455e+05   450.404  413.259
```

**Comment by ambrad** [comment]
Next, I ran the same F-case on cori-knl for two months with PE layout 44x32x4. The two runs are BFB. Again, the timers show no significant difference:
```
>>> master
"CPL:RUN_LOOP"          408 3.987456e+06   2.336051e+06  1659.632 1658.925
"a:CAM_run2"            408 3.988864e+06   3.327829e+05   243.322  226.512
"a:CAM_run3"            408 3.988864e+06   3.910501e+05   284.777  270.367
"a:PAT_remap"           400 1.529820e+07   5.524017e+05   107.498   93.740
"a:SLMM_reconstruct"    400 1.529820e+07   9.744512e+03     1.903    1.164
"a:SLMM_v2x"            400 1.529820e+07   1.176941e+04     2.325    1.575
"a:SLMM_csl"            400 1.529820e+07   2.144449e+05    43.356   36.308
"a:CEDR"                400 1.529820e+07   1.093857e+05    23.750   15.775
"a:CEDR_local"          400 1.529820e+07   4.443359e+04     8.662    7.979
"a:SLMM_bexchV"         400 1.529820e+07   3.209260e+04    10.730    1.399
"a:SLMM vertical remap" 400 1.529820e+07   1.138860e+05    22.099   14.095
"a:CAM_run4"            408 3.988864e+06   3.794200e+04    30.879   26.693
"a:CAM_run1"            408 3.988864e+06   1.295302e+06   970.450  800.166
>>> branch
"CPL:RUN_LOOP"          408 3.987456e+06   2.315361e+06  1644.924 1644.241
"a:CAM_run2"            408 3.988864e+06   3.296248e+05   239.271  227.328
"a:CAM_run3"            408 3.988864e+06   3.842362e+05   281.505  264.694
"a:PAT_remap"           400 1.529820e+07   5.420452e+05   107.326   92.702
"a:SLMM_reconstruct"    400 1.529820e+07   9.754881e+03     1.903    1.165
"a:SLMM_v2x"            400 1.529820e+07   1.173200e+04     2.351    1.601
"a:SLMM_csl"            400 1.529820e+07   2.204616e+05    43.852   37.413
"a:CEDR"                400 1.529820e+07   9.263128e+04    19.392   14.791
"a:CEDR_local"          400 1.529820e+07   5.273915e+04    10.003    9.544
"a:SLMM_bexchV"         400 1.529820e+07   3.319618e+04     9.681    1.399
"a:SLMM vertical remap" 400 1.529820e+07   1.142003e+05    21.991   14.133
"a:CAM_run4"            408 3.988864e+06   3.588538e+04    29.416   25.235
"a:CAM_run1"            408 3.988864e+06   1.293859e+06   969.092  798.800
```

**Comment by ambrad** [comment]
This PR should not be merged until after the v2 WC tag is established.

**Comment by ambrad** [comment]
@bartgol I'm asking for your review, but feel free to ignore large parts of it, probably everything in the compose/ folder. To orient you, the Hommexx dycore entry points are:
- ComposeTransport*.*pp
- GllFvRemap*.*pp
- compose_ut.cpp
- gllfvremap_ut.cpp

**Comment by ambrad** [comment]
Performance results for Hommexx with SL transport are documented [here](https://acme-climate.atlassian.net/wiki/spaces/NGDSA/pages/2516254739/SL+Transport+on+GPU), with data available [here](https://github.com/E3SM-Project/perf-data/tree/main/nhxx-sl-summit-mar2021).

**Comment by ambrad** [comment]
With chrysalis-bfb.cmake, I get this for standalone Homme:
```
The following tests FAILED:
    1 - verifyBaselineResults (Failed)
   32 - thetah-sl-dcmip16_test1pg2 (Failed)
   61 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-sl-ne2-nu3.4e18-ndays1 (Failed)
   62 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-sl-kokkos-ne2-nu3.4e18-ndays1 (Failed)
   79 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-sl-ne6-nu1.25e17-ndays1 (Failed)
   80 - theta-f1-tt10-hvs1-hvst0-r2-qz10-nutopoff-GB-sl-kokkos-ne6-nu1.25e17-ndays1 (Failed)
```
I think thetah-sl-dcmip16_test1pg2 has a memory bug somewhere in the test itself, so it keeps diffing when I touch physgrid or Sl transport. I'll try to find the issue and fix it at some point, outside of this PR.

**Comment by ambrad** [comment]
With chrysalis.cmake, I get this for standalone Homme:
```
The following tests FAILED:
   30 - thetah-sl-test11conv-r1t2-cdr20 (Failed)
   31 - thetah-sl-test11conv-r0t1-cdr30-rrm (Failed)
   32 - thetah-sl-dcmip16_test1pg2 (Failed)
```

File: components/homme/src/share/cxx/prim_cxx_driver_base.F90 [hunk 1]
```diff
@@ -108,6 +114,8 @@ end subroutine prim_init1
 
   subroutine prim_finalize ()
     use prim_driver_base, only: prim_finalize_base=>prim_finalize
+    use control_mod, only: transport_alg
```
Comment (bartgol) [inline components/homme/src/share/cxx/prim_cxx_driver_base.F90:117]: These two seem to be unused.

**Review by bartgol — COMMENTED** [review]
This is a huge PR. I skimmed through most of it, but didn't get into the impl details. I mostly have some questions, nor really change requests.

If you'd like me to get into the impl details, let me know.

**Comment by ambrad** [comment]
On weaver in the BFB build, the new unit tests and the preqx and theta cxx_vs_f90 tests pass.

File: components/homme/cmake/HommeMacros.cmake [hunk 2]
```diff
@@ -130,7 +130,14 @@ macro(createTestExec execName execType macroNP macroNC
...
 
-  TARGET_LINK_LIBRARIES(${execName} timing ${COMPOSE_LIBRARY} ${BLAS_LIBRARIES} ${LAPACK_LIBRARIES})
+  string(FIND ${execType} "kokkos" KOKKOS_SUFFIX_LOC)
+  if (KOKKOS_SUFFIX_LOC EQUAL -1)
+    set (COMPOSE_LIBRARY_TYPE ${COMPOSE_LIBRARY_F90})
+  else ()
+    set (COMPOSE_LIBRARY_TYPE ${COMPOSE_LIBRARY_CPP})
+  endif ()
+
+  TARGET_LINK_LIBRARIES(${execName} timing ${COMPOSE_LIBRARY_TYPE} ${BLAS_LIBRARIES} ${LAPACK_LIBRARIES})
```
[the remark this answers is not in the document]
Comment (ambrad) [reply components/homme/cmake/HommeMacros.cmake:140]: My intent was to have it to expand to nothing if HOMME_ENABLE_COMPOSE is false. If you don't like that pattern, I can write a separate TARGET_LINK_LIBRARIES line for ${COMPOSE_LIBRARY_TYPE}.

File: components/homme/src/share/compose/cedr_caas.cpp [hunk 3]
```diff
@@ -0,0 +1,409 @@
+// COMPOSE version 1.0: Copyright 2018 NTESS. This software is released under
+// the BSD license; see LICENSE in the top-level directory.
+
+#include "cedr_caas.hpp"
+#include "cedr_util.hpp"
+#include "cedr_test_randomized.hpp"
+
+namespace Kokkos {
+struct Real2 {
```
[the remark this answers is not in the document]
Comment (ambrad) [reply components/homme/src/share/compose/cedr_caas.cpp:9]: C++ permits structs to be multiply defined, unlike functions, although the choice of symbol resolution is undefined. But I'll rename Real2 to ComposeReal2 here.

File: components/homme/src/share/compose/compose.hpp [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+#define INCLUDE_COMPOSE_HPP
+
+#ifdef HAVE_CONFIG_H
+# include "config.h.c"
+#endif
+
+#include <Kokkos_Core.hpp>
+
+#ifdef _OPENMP
+# include <omp.h>
```
[the remark this answers is not in the document]
Comment (ambrad) [reply components/homme/src/share/compose/compose.hpp:11]: I use OpenMP library routines.

File: components/homme/src/share/compose/compose_cedr_caas.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+namespace compose {
+
+typedef cedr::Int Int;
+typedef cedr::Real Real;
+
+// On Chrysalis, optimization level above O1 leads to BFB variance w.r.t. PE
+// layout on the WC v2 problem, despite the e3sm_atm_developer and
+// e3sm_atm_integration passing against baselines.
+#ifdef __INTEL_COMPILER
+# pragma intel optimization_level 1
```
[the remark this answers is not in the document]
Comment (ambrad) [reply components/homme/src/share/compose/compose_cedr_caas.cpp:15]: It's in the noise.

File: components/homme/src/share/cxx/prim_cxx_driver_base.F90 (hunk 1 above)
Comment (ambrad) [reply components/homme/src/share/cxx/prim_cxx_driver_base.F90:117]: Thanks; I'll fix that.

File: components/homme/cmake/HommeMacros.cmake (hunk 2 above)
[the remark this answers is not in the document]
Comment (bartgol) [reply components/homme/cmake/HommeMacros.cmake:140]: Nah, it's ok. Perhaps add a comment that clarifies that it expands to nothing if there's no compose.

[the remark this answers is not in the document]
Comment (ambrad) [reply components/homme/cmake/HommeMacros.cmake:144]: Done.

File: components/homme/src/share/compose/cedr_caas.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (ambrad) [reply components/homme/src/share/compose/cedr_caas.cpp]: Done.

**Comment by ambrad** [comment]
@mt5555 , sure, but before I do, what do you think of relying on the repo https://github.com/e3sm-project/perf-data/ for documenting specific performance collection runs? The output, e.g. https://github.com/E3SM-Project/perf-data/blob/main/nhxx-sl-summit-mar2021/data/qsize10/r2-ne1024-nmax4096-qsize10-nnode1024-out.852006, shows the namelist, and the job script template with run configuration details is here: https://github.com/E3SM-Project/perf-data/blob/main/nhxx-sl-summit-mar2021/job.sh.template

**Comment by mt5555** [comment]
Yes, that is a better idea.  In which case we should update the README in homme/test/benchmarks to have a pointer to those other repos.  That probably does not belong in this PR.    I will make those changes in a separate HOMME cleanup PR.

File: components/homme/src/share/cxx/HybridVCoord.cpp [hunk 6]
```diff
@@ -221,11 +221,14 @@ void HybridVCoord::compute_eta ()
...
   auto l_hybrid_bi = hybrid_bi;
+  const auto l_ps0 = ps0;
+  const auto p0 = PhysicalConstants::p0;
+  const auto kappa = PhysicalConstants::kappa;
 
   Kokkos::parallel_for(Kokkos::RangePolicy<ExecSpace>(0,NUM_LEV),
                        KOKKOS_LAMBDA(const int& ilev){
     l_etam(ilev) = l_hybrid_am(ilev) + l_hybrid_bm(ilev);
-    l_exner0(ilev) = pow(l_etam(ilev)*ps0/PhysicalConstants::p0,PhysicalConstants::kappa);
+    l_exner0(ilev) = pow(l_etam(ilev)*l_ps0/p0,kappa);
```
Comment (oksanaguba) [inline components/homme/src/share/cxx/HybridVCoord.cpp:231]: these seems to be triggered randomly, another one is in HyperviscosityFunctorImpl.cpp file, Luca fixed it for a100 builds, there are probably more in the code.

Comment (ambrad) [reply components/homme/src/share/cxx/HybridVCoord.cpp:231]: Agreed. Using a host constant should always fail, but in practice it depends on how the code is compiled, so sometimes these misuses stick around for a while.

Separately, and to be clear, this modification has nothing to do with the main content of the PR; I was just fixing something in Hommexx I found when running cuda-memcheck.

**Comment by ambrad** [comment]
I removed "WIP" from the title because the new postv2.0 label indicates this must wait on the v2.0 tag that will be made by I believe 13 Sep, whereas "WIP" is less clear about what this PR is waiting on.

**Comment by ambrad** [comment]
This PR is ready, but before merging it into next, I want to rerun integration tests against baselines on Chrysalis and Compy since (1) master has advanced a fair bit since I last ran them and (2) I believe a number of tests are changing before the v2 tag is made.

**Comment by ambrad** [comment]
After merging master into this PR branch 14 Sep, e3sm_atm_developer and e3sm_atm_integration pass against baselines on Chrysalis.

**Comment by ambrad** [comment]
I ran more tests again the 14 Sep master.

After the eam-hommexx fixes, this PR passes e3sm_developer against today's baselines on Chrysalis. There was one fail in RUN for SMS_D_Ln5.ne4_oQU240.F2010.chrysalis_intel, with a seg fault in ELM initialization, but that test succeeded when I reran it. For reference, here's the stack trace:
```
70: e3sm.exe           000000000BD99069  piolib_mod_mp_pio         995  piolib_mod.F90
70: e3sm.exe           0000000008D89AFF  ncdio_pio_mp_ncd_         194  ncdio_pio.F90.in
70: e3sm.exe           0000000006BFCF3F  snowsnicarmod_mp_        1503  SnowSnicarMod.F90
70: e3sm.exe           000000000577433C  elm_initializemod         584  elm_initializeMod.F90
70: e3sm.exe           00000000056C2305  lnd_comp_mct_mp_l         273  lnd_comp_mct.F90
70: e3sm.exe           0000000000490A13  component_mod_mp_         257  component_mod.F90
70: e3sm.exe           0000000000439604  cime_comp_mod_mp_        1425  cime_comp_mod.F90
70: e3sm.exe           0000000000487505  MAIN__                    122  cime_driver.F90
```
This possible issue seems orthogonal to this PR.

I ran 3sm_atm_developer and e3sm_atm_integration on Compy with Intel. Here I had to generate my own baselines from master. All tests pass except SMS_D_Ln5.ne4_oQU240.F2010.compy_intel.eam-clubb_sp, which reports
```
93: forrtl: error (72): floating overflow
93: Image              PC                Routine            Line        Source
93: e3sm.exe           000000000B4AF8B4  Unknown               Unknown  Unknown
93: libpthread-2.17.s  00002AEAC58D45D0  Unknown               Unknown  Unknown
93: e3sm.exe           000000000236FBE4  clubb_intr_mp_clu        2067  clubb_intr.F90
93: e3sm.exe           0000000002942A3C  physpkg_mp_tphysb        2460  physpkg.F90
93: e3sm.exe           0000000002911BC0  physpkg_mp_phys_r        1034  physpkg.F90
93: e3sm.exe           00000000008BD096  cam_comp_mp_cam_r         250  cam_comp.F90
93: e3sm.exe           000000000087741D  atm_comp_mct_mp_a         396  atm_comp_mct.F90
93: e3sm.exe           000000000048AF3F  component_mod_mp_         257  component_mod.F90
93: e3sm.exe           00000000004423B3  cime_comp_mod_mp_        2285  cime_comp_mod.F90
93: e3sm.exe           00000000004818F4  MAIN__                    122  cime_driver.F90
93: e3sm.exe           000000000041FB92  Unknown               Unknown  Unknown
93: libc-2.17.so       00002AEAC601D3D5  __libc_start_main     Unknown  Unknown
93: e3sm.exe           000000000041FAA9  Unknown               Unknown  Unknown
````
However, Compy-Intel doesn't run in the nightlies, so I suspect again this failure is orthogonal to this PR and just hasn't been noticed before.

Next I'll run Compy-PGI.

**Review by mt5555 — APPROVED** [review]

**Comment by oksanaguba** [comment]
I am building  a branch off this branch on summit with homme_enable_compose=false and getting the message below. turning on compose worked. not sure we would even want option without compose.
```
[100%] Linking Fortran executable theta-nlev128-kokkos
CMakeFiles/theta-nlev128-kokkos.dir/__/__/src/share/cxx/ComposeTransportImplTest2D.cpp.o: In function `operator()':
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplTest2D.cpp:25: undefined reference to `compose::test::InitialCondition::init(compose::test::InitialCondition::Shape, int, double const*, double const*, double*)'
CMakeFiles/theta-nlev128-kokkos.dir/__/__/src/share/cxx/ComposeTransportImplGeneral.cpp.o: In function `Homme::ComposeTransportImpl::run(Homme::TimeLevel const&, double)':
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplGeneral.cpp:176: undefined reference to `homme::compose::advect(int, int, int)'
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplGeneral.cpp:187: undefined reference to `homme::compose::set_dp3d_np1(int)'
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplGeneral.cpp:190: undefined reference to `homme::compose::property_preserve_global()'
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplGeneral.cpp:243: undefined reference to `homme::compose::property_preserve_check()'
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplGeneral.cpp:195: undefined reference to `homme::compose::property_preserve_local(int)'
CMakeFiles/theta-nlev128-kokkos.dir/__/__/src/share/cxx/ComposeTransportImplGeneral.cpp.o: In function `Homme::ComposeTransportImpl::reset(Homme::SimulationParams const&)':
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/ComposeTransportImplGeneral.cpp:62: undefined reference to `homme::compose::set_views(Kokkos::View<double***, Kokkos::LayoutRight, Kokkos::Cuda> const&, Kokkos::View<double****, Kokkos::LayoutRight, Kokkos::Cuda> const&, Kokkos::View<double*****, Kokkos::LayoutRight, Kokkos::Cuda> const&, Kokkos::View<double******, Kokkos::LayoutRight, Kokkos::Cuda> const&, Kokkos::View<double*****, Kokkos::LayoutRight, Kokkos::Cuda> const&, Kokkos::View<double*****, Kokkos::LayoutRight, Kokkos::Cuda> const&)'
collect2: error: ld returned 1 exit status
```

**Comment by oksanaguba** [comment]
I am also getting these,
```
/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(486): warning: Implicit capture of 'this' in extended lambda expression

/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(492): warning: Implicit capture of 'this' in extended lambda expression

/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(510): warning: Implicit capture of 'this' in extended lambda expression

/ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(516): warning: Implicit capture of 'this' in extended lambda expression

```
which point to vars m_tu_ne and m_tu_ne_ntr?

**Comment by bartgol** [comment]
> I am building a branch off this branch on summit with homme_enable_compose=false and getting the message below. turning on compose worked. not sure we would even want option without compose.

If I understand correctly, @ambrad's idea is that homme_enable_compose=false should no longer be supported., but he made it so that the user get a compilation error if set to false, so they get the warning that it's time to change their settings.

**Comment by ambrad** [comment]
@onguba re: `homme_enable_compose=false`, what Luca wrote is correct: It's not meant to be supported any longer, but I left it there for now in case for some reason someone ever has to build theta-l with only a Fortran compiler available.

**Comment by bartgol** [comment]
> I am also getting these,
> 
> ```
> /ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(486): warning: Implicit capture of 'this' in extended lambda expression
> 
> /ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(492): warning: Implicit capture of 'this' in extended lambda expression
> 
> /ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(510): warning: Implicit capture of 'this' in extended lambda expression
> 
> /ccs/home/onguba/acme-b2/components/homme/src/share/cxx/RemapFunctor.hpp(516): warning: Implicit capture of 'this' in extended lambda expression
> ```
> 
> which point to vars m_tu_ne and m_tu_ne_ntr?

Ah, yes, good catch. Those are legit errors. The fix should be the usual "create a local copy before the ||for".

**Comment by ambrad** [comment]
Re: the implicit capture of `this`: Thanks for the warning messages. I see why that hasn't affected my builds before: the variable is used only if HOMMEXX_CUDA_SHARE_BUFFER is enabled, which we don't enable unless necessary. I'll fix those lines. Should I add a commit to this PR or make a new one?

**Comment by oksanaguba** [comment]
Separate PR is best, thanks. I am not sure why i would have that cuda buffer variable on, i just ran a simple cmake line. i will check.

**Comment by ambrad** [comment]
Re: HOMMEXX_CUDA_SHARE_BUFFER, what I mean is that this invalid access should matter only if it's enabled. The warning itself I haven't seen before, but the newest Cuda version might be providing that. I just checked on weaver and don't get the warning, and cuda-memcheck is clean if HOMMEXX_CUDA_SHARE_BUFFER is off.

**Comment by bartgol** [comment]
@ambrad do you want to put that fix in the forcing functor fix PR? That one is yet to be merged in next, so we could still piggy back on that. As you said, it's not urgent, so either way is fine. But since we have a PR already up...

**Comment by ambrad** [comment]
Thanks Luca. @oksanaguba, what is your preference? I can add a commit to Luca's ForcingFunctor PR, or I can make a separate PR, whichever is easier for you.

I fixed a few more of these cases. After the fixes and with HOMMEXX_CUDA_SHARE_BUFFER enabled, cuda-memcheck is clean for compose_ut and gllfvremap_ut, just like it already was with HOMMEXX_CUDA_SHARE_BUFFER disabled.

**Comment by ndkeen** [comment]
I also see those warnings (and others) on Perlmutter.

**Comment by ambrad** [comment]
@ndkeen, thanks. Some warnings are unavoidable. They have to do with some of SL's interface machinery to support both F90 Homme and C++ Hommexx. The ones that I know about that are innocuous are:
* dynamic initialization in unreachable code (due to switches for host vs device exe);
* statement is unreachable (same);
* compose_homme.hpp(165): warning: calling a __host__ function from a __host__ __device__ function is not allowed (due to supporting Homme F90 pointers in place as well as Hommexx device pointers).

Do you see warnings in the SL code other than these?

**Comment by ndkeen** [comment]
`theta-l_kokkos/cxx/ElementOps.hpp(17): warning: __host__ annotation is ignored on a function("ElementOps") that is explicitly defaulted on its first declaration` ?

`homme/src/share/cxx/utilities/BfbUtils.hpp", line 40: warning: integer conversion resulted in a change of sign`

`externals/kokkos/core/src/Cuda/Kokkos_Cuda_Half.hpp", line 318: warning: variable "val_ref" was set but never used`

And of course there were the actual errors from nvidia where I need to make a local `const int`

**Comment by ambrad** [comment]
Thanks @ndkeen. I added a commit to PR #4571 to fix the ElementOps.hpp and BfbUtils.hpp ones. The warnings and the fixes don't affect runtime behavior.

**Comment by oksanaguba** [comment]
With F executable and compose on, exec does not build,
```
/ccs/home/onguba/acme-b2/components/homme/src/share/compose/compose_cedr_sl_run_local.cpp: In function 'void homme::sl::run_local(homme::CDR<MT>&, CDRT*, const homme::sl::Data&, homme::Real*, const Real*, homme::Int, homme::Int, bool, homme::Int)':
/ccs/home/onguba/acme-b2/components/homme/src/share/compose/compose_cedr_sl_run_local.cpp:200:47: error: no matching function for call to 'subview_ie(homme::Int&, const homme::HommeFormatArray<const double, 2>&)'
  200 |     const auto spheremp1 = subview_ie(ie, spheremp);
      |       
...
```
and other errors.

**Comment by ambrad** [comment]
@oksanaguba what machine and configuration is this?

**Comment by oksanaguba** [comment]
Still summit, with summit-gpumpi.cmake file (tweaked for netcdf) and SET (HOMMEXX_BFB_TESTING TRUE CACHE BOOL "") .

module list:
```
bash-4.4$ module list

Currently Loaded Modules:
  1) gcc/9.1.0                        5) cuda/11.0.3            9) netcdf-fortran/4.4.5
  2) spectrum-mpi/10.4.0.3-20210112   6) cmake/3.18.4          10) openblas/0.3.5
  3) nsight-compute/2021.2.1          7) python/3.7-anaconda3
  4) nsight-systems/2021.3.1.54       8) netcdf-c/4.8.0

export PAMI_ENABLE_STRIPING=1
export PAMI_IBV_ADAPTER_AFFINITY=1
export PAMI_IBV_DEVICE_NAME="mlx5_0:1,mlx5_3:1"
export PAMI_IBV_DEVICE_NAME_1="mlx5_3:1,mlx5_0:1"
export PAMI_CUDA_AWARE_THRESH=320000

 ```

**Comment by ambrad** [comment]
Right now I can't reproduce. Here is my env, Homme config, and build line. What should I do differently to reproduce the issue you see? Thanks.

```
$ cat source.sh
module purge
module load gcc/9.1.0 cuda/11.0.3 netcdf-fortran/4.4.5 spectrum-mpi/10.4.0.3-20210112 cmake/3.18.4 openblas/0.3.5 nsight-compute/2021.2.1 python/3.7-anaconda3 nsight-systems/2021.3.1.54 netcdf-c/4.8.0

export PAMI_ENABLE_STRIPING=1
export PAMI_IBV_ADAPTER_AFFINITY=1
export PAMI_IBV_DEVICE_NAME="mlx5_0:1,mlx5_3:1"
export PAMI_IBV_DEVICE_NAME_1="mlx5_3:1,mlx5_0:1"
export PAMI_CUDA_AWARE_THRESH=320000

$ cat config.sh
rm -rf CMake*
cmake \
    -C /ccs/home/ambradl/repo/e3sm/components/homme/cmake/machineFiles/summit-gpumpi.cmake \
    -D HOMMEXX_BFB_TESTING=ON \
    -D CMAKE_CXX_COMPILER="/ccs/home/ambradl/repo/e3sm/externals/kokkos/bin/nvcc_wrapper" \
    -D Kokkos_ENABLE_CUDA_ARCH_LINKING=OFF \
    /ccs/home/ambradl/repo/e3sm/components/homme

$ make -j4 theta-nlev128
...
[100%] Built target theta-nlev128
```

**Comment by oksanaguba** [comment]
Andrew, I still see warnings about this->, like in RemapFunctor.hpp .

**Comment by ambrad** [comment]
Oh, I see, in cedr_qlt.cpp, right? (By the way, only within the past month or two did we have access to a Cuda version that provides these warnings, which is why previously instances of this issue tended to occur frequently.) We can ignore the cedr_qlt.cpp one, as QLT isn't used in any simulations that matter. I'll fix it in a future PR.

But do you really still see one in RemapFunctor.hpp? I did a clean build of compose_ut on Summit, using the environment and configuration I documented above (with HOMME_ENABLE_COMPOSE=ON), and with the branch in this current state (last commit: `26f4cb469a Hommexx: Fix limiters_ut in non-BFB standalone test.`) and see only a `this->` warning for cedr_qlt.cpp.

**Comment by rljacob** [comment]
@ambrad How is this PR different from https://github.com/E3SM-Project/E3SM/pull/2896 ?  They both say "Add SL transport".

**Pull request merged**